--- ./missing.h
+++ ./missing.h
@@ -48,12 +48,6 @@
 #define FD_TO_CLOCKID(fd)	((~(clockid_t) (fd) << 3) | CLOCKFD)
 #define CLOCKID_TO_FD(clk)	((unsigned int) ~((clk) >> 3))
 
-#ifndef HAVE_ONESTEP_SYNC
-enum _missing_hwtstamp_tx_types {
-	HWTSTAMP_TX_ONESTEP_SYNC = 2,
-};
-#endif
-
 #ifndef SIOCGHWTSTAMP
 #define SIOCGHWTSTAMP 0x89b1
 #endif
@@ -62,12 +56,6 @@
 #define SO_SELECT_ERR_QUEUE 45
 #endif
 
-#ifndef HAVE_CLOCK_ADJTIME
-static inline int clock_adjtime(clockid_t id, struct timex *tx)
-{
-	return syscall(__NR_clock_adjtime, id, tx);
-}
-#endif
 
 #ifndef __uClinux__
 
